File: courses/ce594/hw3/src/fe.js

var math = require( 'mathjs' );

var x_left = 0;
var x_right = 1;
var left_boundary_value = 0;
var right_boundary_value = 0;
var num_elements = 8;
var num_element_nodes = 2;

// Create mesh
var mesh = Mesh(
    x_left,
    x_right,
    left_boundary_value,
    right_boundary_value,
    num_elements,
    num_element_nodes
);

// Solve 1d Head equation
var area = 0.1;
var k = 0.5;
var source = 5;

Heat_1d( area, k, source, mesh );


function Mesh ( xl, xr, bl, br, num_elements, num_element_nodes ) {

    // Create the IEN matrix
    var IEN = math.zeros( num_element_nodes, num_elements );
    var node_number = 1;

    for ( var element=0; element<num_elements; ++element ) {

        for ( var node=0; node<num_element_nodes; ++node ) {

            IEN.set( [ node, element ], node_number++ );

        }

        node_number--;

    }

    // Create the ID matrix
    var num_nodes = num_elements * ( num_element_nodes - 1 ) + 1;
    var ID = [];

    for ( node=0; node<num_nodes; ++node ) {
        ID.push( 0 );
    }
    for ( node=1; node<num_nodes-1; ++node ) {
        ID[node] = node;
    }


    // Create the coordinates
    var x_coordinates = [];
    var interval = ( xr - xl ) / ( num_nodes - 1);

    for ( node=0; node<num_nodes; ++node ) {
        x_coordinates.push( xl + node*interval );
    }


    // Return mesh object
    return {
        x_coordinates: x_coordinates,
        ID: ID,
        IEN: IEN,
        left_boundary: bl,
        right_boundary: br,
        degrees_of_freedom: num_nodes-2,
        num_nodes: num_nodes,
        num_elements: num_elements,
        num_element_nodes: num_element_nodes
    }


}

function Heat_1d ( area, k, source, mesh ) {

    // Set up one point integration
    var num_integration_points = 1;
    var x_integration = 0;
    var w_integration = 2;
    var Nt = [ 0.5, 0.5 ];
    var dNt = [ -0.5, 0.5 ];

    // Set up global arrays
    var K = math.zeros( mesh.degrees_of_freedom, mesh.degrees_of_freedom );
    var F = math.zeros( 1, mesh.degrees_of_freedom );

    // Element loop
    for ( var element=0; element<mesh.num_elements; ++element ) {

        console.log( 'Element', element + 1 );

        var ke = math.zeros( mesh.num_element_nodes, mesh.num_element_nodes );
        var fe = math.zeros( 1, mesh.num_element_nodes );

        var x_left = mesh.x_coordinates[ mesh.IEN.get( [0, element] ) ];
        var x_right = mesh.x_coordinates[ mesh.IEN.get( [1, element] ) ];
        var J = jacobian( x_left, x_right );

        // Integration loop
        for ( var i=0; i<num_integration_points; ++i ) {

            console.log( '\tIntegration point', i+1 );

            var dNdx = [ dNt[0] / J, dNt[1] / J ];
            var N = [ Nt[0], Nt[1] ];

            // Node loops
            for ( var a=0; a<mesh.num_element_nodes; ++a ) {

                console.log( '\t\tNode', a+1 );

                fe.set( [1, a], N[a] * source * J * w_integration );

                for ( var b=0; b<mesh.num_element_nodes; ++b ) {

                    console.log( '\t\t\tNode', b+1 );
                    ke.set( [b, a], dNdx[a] * area * k * dNdx[b] * J * w_integration );

                }

            }
        }

        // Update K and F with contributions from ke and fe
        for ( a=0; a<mesh.num_element_nodes; ++a ) {

            console.log( '\tUpdate K and F' );

            var A = mesh.IEN.get( [a, element] )-1;
            var P = mesh.ID[ A ]-1;

            if ( P > -1 ) {

                console.log( '\t\tUpdate F, Node', a+1 );
                F.set( [0, P], F.get( [0, P] ) + fe.get( [0,a] ) );

                for ( b=0; b<mesh.num_element_nodes; ++b ) {

                    var B = mesh.IEN.get( [b, element] )-1;
                    var Q = mesh.ID[ B ]-1;

                    if ( Q > -1 ) {

                        console.log( '\t\t\tUpdate K, Node', b+1 );
                        K.set( [Q,P], K.get( [Q,P] ) + ke.get( [b,a] ))

                    }

                }

            }

        }

    }

    console.log( K );
    console.log( F );

}

function jacobian ( xl, xr ) {
    return 0.5 * ( xr - xl );
}

File: courses/ce594/hw3/src/fe.py
import numpy as np
import matplotlib.pyplot as plt

# Mesh
num_elements = 8
num_element_nodes = 2
num_nodes = num_elements * ( num_element_nodes - 1 ) + 1
x_coord = [ float(i)/num_elements for i in range(num_nodes) ]

# Use functions for b.c.'s to make things a little quicker
def h ( el, el_node ):
    if el == num_elements-1 and el_node == 1:
        return 25
    return 0

def u ( el, el_node ):
    if el == 0 and el_node == 0:
        return 10
    return 0

# Inputs
area = 0.1
k = 0.5
def s ( _x ):
    return 5 #*_x

# Integration
num_int_points = 1
x_int = 1
w_int = 2

Ntilde = [ 0.5, 0.5 ]
dNtilde = [ -0.5, 0.5 ]

# Global arrays
K = np.zeros( ( num_nodes-1, num_nodes-1 ) )
F = np.zeros( ( num_nodes-1, 1 ) )

# Global node number given an element index (i.e. starting at element zero)
# and an element node index (i.e. also starting at zero)
#   e.g. IEN(0,0) - element 0 and node 0 within that element
def IEN ( _element, _element_node ):

    return _element+_element_node

# Get index into K and F arrays given a node number
# Returns -1 if the node is not in K and F
def ID ( A ):

    # return A
    return -1 if A == 0 else A-1


# Element loop
for element in range( num_elements ):

    ke = np.zeros( ( num_element_nodes, num_element_nodes ) )
    fe = np.zeros( ( num_element_nodes, 1 ) )

    i_left = IEN( element, 0 )
    i_right = IEN( element, 1 )

    x_element = [ x_coord[ i_left ], x_coord[ i_right ] ]

    J = 0.5 * ( x_element[1] - x_element[0] )

    # Integration loop
    for i in range( num_int_points ):

        N = [ Nt for Nt in Ntilde ]
        dNdx = [ dNt/J for dNt in dNtilde ]

        # Node loop
        for row in range( num_element_nodes ):

            fe[ row, 0 ] = ( N[ row ] * s( x_element[ row ] ) * J * w_int ) + \
                           ( area * N[ row ] * h( element, row ) )

            for col in range( num_element_nodes ):

                ke[ row, col ] = dNdx[ row ] * area * k * dNdx[ col ] * J * w_int
                fe[ row, 0 ] -= ke[ row, col ] * u( element, row )

    print fe
    print ke

    # Update globals with contributions from local
    for row in range( num_element_nodes ):

        row_node = IEN( element, row )
        ROW = ID( row_node )

        if ROW != -1:

            F[ ROW, 0 ] += fe[ row, 0 ]

            for col in range( num_element_nodes ):

                col_node = IEN( element, col )
                COL = ID( col_node )

                if COL != -1:

                    K[ ROW, COL ] += ke[ row, col ]

d = np.linalg.solve( K, F )
print d
x = x_coord
y = d[:,0]

y = np.insert( y, 0, 0 )
# y = np.append( y, 0 )

#plt.plot( x, y )
#plt.show()